# Script Contents
### -Import Libraries

### -Import Data

### -Create a Mixed-Type Column in a Test Dataframe

### -Find Missing Values

### -Find Duplicates

### -Export Data

### -Task 4.5

# Importing Libraries

In [3]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# Importing Data

In [6]:
# Create a path for my pain project folder
path = r"C:\Users\jacks\Instacart Basket Analysis 081424"

In [8]:
# Import the order data from Prepared Data and the product data from Original Data
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# Create a Mixed-Type Column in a Test Dataframe

In [11]:
# Create a dataframe
df_test = pd.DataFrame()

In [13]:
# Create a mixed type columm
df_test['mix'] = ['a', 'b', 1, True]

In [15]:
# Check out the test dataframe
df_test.head()

mix
0     a
1     b
2     1
3  True

In [23]:
# Check for mixed types
for col in df_test.columns.tolist():
    weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_test[weird]) > 0:
        print(col)

mix


In [29]:
# Convert column's datatype into string only
df_test['mix'] = df_test['mix'].astype('str')

# Finding Missing Values

In [32]:
# Find missing observations in the product data and add up the number of missing observations
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [34]:
# Create a new df that contains only the missing values from df_prods
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [36]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [38]:
# Reminder of the number of rows and columns in df_prods
df_prods.shape

(49693, 5)

In [44]:
# Create a new product df that doesn't contain missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [46]:
df_prods_clean.shape

(49677, 5)

# Duplicates

In [51]:
# Find full duplicates in df_prods_clean by creating a subset
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [53]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [59]:
# Checking to see how the duplicated function works. Does it provide the second instance of the row's data to occur in the df?
df_prods_clean[df_prods_clean['product_name'] == 'Ranger IPA']

product_id product_name  aisle_id  department_id  prices
18458       18458   Ranger IPA        27              5     9.2
18459       18458   Ranger IPA        27              5     9.2

In [61]:
# Drop duplicates in a new df
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [65]:
# Check shape of new df. Should have less than 49,677 rows
df_prods_clean_no_dups.shape

(49672, 5)

# Export Product Data

In [70]:
# Exporting to Prepared Data folder
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task 4.5

## 02. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [75]:
# Describe df_ords
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_prior_order  
count        3.421083e+06       3.421083e+06            3.214874e+06  
mean         2.776219e+00       1.345202e+01            1.111484e+01  
std          2.046829e+00       4.226088e+00            9.206737e+00  
min          0.000000e+00       0.000000e+00            0.000000e+00  
25%          1.000000e+00       1.000000e+01            4.000000e+00  
50%          3.000000e+00       1.300000e+01            7.000000e+00  
75%          5.000000e+00       1.600000e+01            1.500000e+01  
max          6.000000e+00       2.300000e+01            3.000000e+01

#### The count of days_since_prior_order is lower than all other columns. This tells me there could be missing values in this column. The order_number column has a min value of 1 and a max value of 100. Does this mean the most orders any user_id has placed is 100? Maybe the data is capped at 100 orders per user_id. The order_hour_of_day column ranges from 0-23. I assume 0 in this column is 12am. 

## 03. Check for mixed-type data in your df_ords dataframe.

In [80]:
# Check for mixed types
for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print(col)

#### Since no columns printed, there does not appear to be any mixed-type data

## 05. Run a check for missing values in your df_ords dataframe..

In [86]:
# Find missing observations in the orders data and add up the number of missing observations
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

#### There are 206,209 missing values in the days_since_prior_order column. Looking back at the description of df_ords, the mean value of days_since_prior_order is 11.11, the max is 30, the median is 7, and the upper quartile is 15. Since the majority of the observations are 15 or less, I think the mean of 11.11 is being pulled higher than the median of 7 due to the high values between 16-30 in the last quartile. For this reason, I believe it makes more sense to replace the missing values with the median of 7 instead of the mean of 11.11. Dropping the missing values does not seem reasonable because of the number of missing values.

## .06 Address the missing values using an appropriate method.

In [103]:
# Double-check the median of the column
df_ords.median()

Unnamed: 0                1710541.0
order_id                  1710542.0
user_id                    102689.0
order_number                   11.0
orders_day_of_week              3.0
order_hour_of_day              13.0
days_since_prior_order          7.0
dtype: float64

In [119]:
# Fill missing values in df_ords days_since...column with median
df_ords.fillna({'days_since_prior_order': 7}, inplace=True)

In [121]:
# Check to see if there are any missing values in updated df
df_ords.isnull().sum()

Unnamed: 0                0
order_id                  0
user_id                   0
order_number              0
orders_day_of_week        0
order_hour_of_day         0
days_since_prior_order    0
dtype: int64

#### I chose to repalce missing values in the days_since_prior_order column with the columns median, 7, since the median appeared to be a more accurate representation of the columns 'middle' than the mean of 11.11, as I explained in the markdown after finding the missing values.

## 07. Run a check for duplicate values in your df_ords data.

In [123]:
# Find full duplicates in df_ords_clean by creating a subset
df_ords_dups = df_ords[df_ords.duplicated()]

In [125]:
# Check to see the results of the subset
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

## .08 Address the duplicates using an appropriate method.

#### There were no duplicates to address. If there were full duplicates, I'd drop them. 

## .09 Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [130]:
# Cleaned df_prods data already exported. 
# Exporting order data to Prepared Data folder
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))